docs/architecture/adr-011-light-client-crates-extraction.md

```diff
@@ -0,0 +1,375 @@
+    /// consensus state associated with the client that corresponds to a given
+    /// height.
+    #[prost(message, optional, tag = "2")]
+    pub consensus_state: ::core::option::Option<super::super::super::super::google::protobuf::Any>,
+    /* ... */
+}
+```
+
+Ideally, the domain type would contain validated `ClientState` and `ConsensusState` types but this is not possible
+anymore because the `ibc` crate cannot depend on the light client crates. It is therefore proposed that domain types
+continue to use the `Any` type for such fields and defer the validation to the handlers.
+
+### Light client specific code
+
+A sizeable amount of light client specific code exists in the `ibc` crate exists today that is only used by `hermes`.
+For e.g. helper functions that extract `IbcEvent`s from `tendermint::abci::Event`s (i.e. `from_tx_response_event()`).
+
+There are cases where core types depend on light client types, for e.g. `ics02_client::Error`
+uses `ics07_tendermint::Error`s in some of its variants.
+
+Additionally, some of that code might be helpful for host implementations, for e.g. event conversions
+to/from `tendermint::abci::Event` are required for `deliverTx` implementation.
+
+It is recommended that most client specific code is moved either to the `ibc-relayer` crate or the light client crates
+and issues of this kind be dealt with on a case-by-case basis.
+
+### Light client registry
+
+With the proposals in this ADR, the `ibc` crate would be light client agnostic, however, the host implementation must
+still be aware of all light clients that it wishes to support. For e.g., after an upgrade a blockchain must be able to
+deserialize types from the persistent store.
+Furthermore, protobuf has emerged as the canonical serialization scheme for IBC, and IBC's message definitions usually
+serialize light client types using the `google::protobuf::Any` type where the `type_url` is accepted to uniquely
+represent specific light client types, although this has not been standardized yet.
+It is proposed to standardize this and provide a `ibc-client-registry` crate which standardizes this and collects all
+known light client implementations. This crate must provide one deserialization entry-point function per light client
+type ->
+
+```rust
+use ibc_proto::google::protobuf::Any;
+
+pub const TENDERMINT_HEADER_TYPE_URL: &str = "/ibc.lightclients.tendermint.v1.Header";
+
+fn decode_header(any_header: Any) -> Result<Box<dyn Header>, Error> {
+    match any_header.type_url {
+        TENDERMINT_HEADER_TYPE_URL => TendermintHeader::try_from(any_header).map_err(|e| Error::decode_header(e)),
+        /* ... */
+        _ => Err(Error::unsupported_client())
+    }
+}
+```
+
+This way we have a standardized list of supported light clients so that host implementations would not have to directly
+import every single light client crate that they wish to support. The crate could feature gate light client support to
```

Comment:
How would light client versioning work in this system?

Given a fixed type url, say `"/ibc.lightclients.tendermint.v1.Header"`, the `tendermint-light-client` crate could change its domain type from version `1.0` to `1.1`. Which version would the registry use? What if the registry uses `1.0`, and my chain uses `1.1`?

I think we need to carefully consider this issue, and then decide whether or not the registry is a good idea, and if it would end up causing more headaches than it solves.

Reply:
I agree this needs careful consideration and we must ask ourselves questions such as these. :+1:

> Which version would the registry use? What if the registry uses 1.0, and my chain uses 1.1?

For the specific problem you described above, I think one of the benefits of having a client registry is that host implementations will not have to import or deal with light client crates/types directly, they will only depend on the light client crates via the registry. Furthermore, both the host and the module code will only ever access light client types from behind a `&dyn Trait` (if they use the registry), so changes to the concrete type would not have any adverse effects (AFAICT). 

The registry can be thought of as hosting the `Any*` enums that we currently have in our codebase. Interestingly, this is indeed one of the ways in which such a registry can be implemented. So usability wise not much should change for the host implementations. Note that the versioning of light client crates does not have an immediate effect on the versioning scheme of the registry crate because the registry never exposes the concrete light client types. We might need to decide on a versioning scheme for the registry crate, but I don't think that will be too difficult. 

To me, the fact that there isn't a canonical mapping of a `type_url` <> `ClientType` is a clear sign that we need such a registry. 
Would be nice to get more input from the team here. cc: @romac @mzabaluev @soareschen

Reply:
As discussed, I don't think this is an issue anymore. Even if the chain implementing the keepers and readers uses `tendermint-light-client` crate on a different version than whatever the registry uses, Rust sees the types coming from them as separate. For example,
```rust
// The tendermint-light-client crate as declared in the chain's Cargo.toml
use tendermint_light_client::Header;

// even if `light_client_registry` crate re-exports the tendermint_light_client
// crate, then any type is exports (here, `TmHeader`) is a different type from
// `Header`
use light_client_registry::tendermint_light_client::Header as TmHeader;
```
Ultimately, the registry can choose to update its dependencies freely without fear of breaking code that depends on it.